core/src/main/csharp/ch/cyberduck/core/CredentialManagerPasswordStore.cs

```diff
@@ -49,30 +48,178 @@ public override void addPassword(Scheme scheme, int port, string hostName, strin
         }
 
+        public override string findPrivateKeyPassphrase(Host bookmark)
+        {
+            if (logger.isInfoEnabled())
+            {
+                logger.info(string.Format("Fetching private key passphrase from keychain for {0}", bookmark));
+            }
+            var target = ToUri(bookmark);
+            var cred = WinCredentialManager.GetCredentials(target.AbsoluteUri);
+            if (cred.Attributes.TryGetValue("Private Key Passphrase", out var passphrase)
+                && !string.IsNullOrWhiteSpace(passphrase))
+            {
+                return passphrase;
+            }
+            return base.findPrivateKeyPassphrase(bookmark);
+        }
+
+        public override string getPassword(string serviceName, string user) => WinCredentialManager.GetCredentials($"{serviceName} - {user}").Password;
+
         public override string getPassword(Scheme scheme, int port, string hostName, string user)
         {
             var hostUrl = hostUrlProvider.get(scheme, port, user, hostName, string.Empty);
             return WinCredentialManager.GetCredentials(hostUrl).Password;
         }
 
-        public override void deletePassword(String serviceName, String user)
+        public override void save(Host bookmark)
         {
-            WinCredentialManager.RemoveCredentials($"{serviceName} - {user}");
+            if (string.IsNullOrWhiteSpace(bookmark.getHostname()))
+            {
+                logger.warn("No hostname given.");
+                return;
+            }
+            if (logger.isInfoEnabled())
+            {
+                logger.info(string.Format("Add password for bookmark {0}", bookmark));
+            }
+            var target = ToUri(bookmark);
+            var protocol = bookmark.getProtocol();
+            var credential = bookmark.getCredentials();
+
+            var winCred = new WindowsCredentialManagerCredential(
+                credential.getUsername(), credential.getPassword(),
+                CRED_TYPE_GENERIC, 0, CRED_PERSIST_ENTERPRISE);
+
+            if (protocol.isTokenConfigurable())
+            {
+                winCred.Attributes["Token"] = credential.getToken();
+            }
+            if (protocol.isOAuthConfigurable())
+            {
+                winCred.Attributes["OAuth Access Token"] = credential.getOauth().getAccessToken();
+                winCred.Attributes["OAuth Refresh Token"] = credential.getOauth().getRefreshToken();
+                if (credential.getOauth().getExpiryInMilliseconds() != null)
+                {
+                    winCred.Attributes["OAuth Expiry"] = credential.getOauth().getExpiryInMilliseconds().longValue().ToString();
+                }
+            }
+            if (protocol.isPrivateKeyConfigurable())
+            {
+                winCred.Attributes["Private Key Passphrase"] = credential.getIdentityPassphrase();
+            }
+            if (credential.isPasswordAuthentication())
```

Comment:
Need to check here.
If password authentication is required, then if either username or password is empty the whole credential is not saved.

Can private key configurable be used with password authentication, and username is empty?

Reply:
Should match logic in `DefaultHostPasswordStore` to make sure we have no regressions.

Reply:
DefaultHostPasswordStore runs separate addPassword-calls for these conditions.

CredentialManagerPasswordStore now only performs one single commit with all fields combined.